File: complex/src/main/java/net/fwitz/math/plot/in3d/Triangulator.java
package net.fwitz.math.plot.in3d;

import net.fwitz.math.complex.Complex;

import java.util.Arrays;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Triangulator {
    private final Vertex[][] points;
    private final Triangle[][] triangles;

    public Triangulator(int width, int height) {
        this.points = new Vertex[height][];
        for (int y = 0; y < height; y++) {
            final double yd = y;
            points[y] = IntStream.range(0, width)
                    .mapToObj(x -> new Vertex(x, yd, 0.0, 0))
                    .toArray(Vertex[]::new);
        }
        this.triangles = new Triangle[height - 1][];
        for (int y = 0; y < height - 1; y++) {
            refreshTriangles(y);
        }
    }

    public void addRow(int y, Complex[] values, int[] colors) {
        points[y] = IntStream.range(0, colors.length)
                .mapToObj(x -> new Vertex(x, y, values[x].abs(), colors[x]))
                .toArray(Vertex[]::new);
        if (y > 0) {
            refreshTriangles(y - 1);
        }
        if (y < points.length - 1) {
            refreshTriangles(y);
        }
    }

    private void refreshTriangles(int y) {
        Vertex[] above = points[y];
        Vertex[] below = points[y + 1];
        triangles[y] = IntStream.range(0, above.length)
                .mapToObj(x -> Stream.of(
                    new Triangle(above[x], below[x], above[x + 1]),
                    new Triangle(above[x + 1], below[x], below[x + 1])))
                .flatMap(s -> s)
                .toArray(Triangle[]::new);
    }

    public Stream<Triangle> triangles() {
        return Arrays.stream(triangles).flatMap(Arrays::stream);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/in3d/Triangulator.kt
package net.fwitz.math.plot.in3d

class Triangulator(width: Int, height: Int) {
    /*
    private val points: Array<Array<Vertex>>
    private val triangles: Array<Array<Triangle>>
    fun addRow(y: Int, values: Array<Complex>, colors: IntArray) {
        points[y] = IntStream.range(0, colors.size)
            .mapToObj<Vertex>(IntFunction<Vertex> { x: Int ->
                Vertex(
                    x.toDouble(), y.toDouble(), values[x].abs(), colors[x]
                )
            })
            .toArray<Vertex>(IntFunction<Array<Vertex>> { _Dummy_.__Array__() })
        if (y > 0) {
            refreshTriangles(y - 1)
        }
        if (y < points.size - 1) {
            refreshTriangles(y)
        }
    }

    private fun refreshTriangles(y: Int) {
        val above = points[y]
        val below = points[y + 1]
        triangles[y] = IntStream.range(0, above.size)
            .mapToObj<Stream<Triangle>>(IntFunction<Stream<Triangle>> { x: Int ->
                Stream.of(
                    Triangle(above[x], below[x], above[x + 1]),
                    Triangle(above[x + 1], below[x], below[x + 1])
                )
            })
            .flatMap<Triangle>(Function<Stream<Triangle?>, Stream<out Triangle?>?> { s: Stream<Triangle?>? -> s })
            .toArray<Triangle>(IntFunction<Array<Triangle>> { _Dummy_.__Array__() })
    }

    fun triangles(): Stream<Triangle> {
        return Arrays.stream<Array<Triangle>>(triangles)
            .flatMap<Triangle>(Function<Array<Triangle?>, Stream<out Triangle>> { array: Array<Triangle?>? ->
                Arrays.stream(array)
            })
    }

    init {
        points = arrayOfNulls(height)
        for (y in 0 until height) {
            val yd = y.toDouble()
            points[y] = IntStream.range(0, width)
                .mapToObj<Vertex>(IntFunction<Vertex> { x: Int ->
                    Vertex(
                        x.toDouble(), yd, 0.0, 0
                    )
                })
                .toArray<Vertex>(IntFunction<Array<Vertex>> { _Dummy_.__Array__() })
        }
        triangles = arrayOfNulls(height - 1)
        for (y in 0 until height - 1) {
            refreshTriangles(y)
        }
    }
    */
}